youtube_dl/extractor/pandatv.py

```diff
@@ -0,0 +1,87 @@
+# coding: utf-8
+from __future__ import unicode_literals
+
+from .common import InfoExtractor
+from ..utils import ExtractorError, qualities
+from ..compat import (compat_str, compat_basestring)
+
+
+class PandaTVIE(InfoExtractor):
+    IE_DESC = '熊猫TV'
+    _VALID_URL = r'http://(?:www\.)?panda\.tv/(?P<id>[0-9]+)'
+    _TESTS = [{
+        'url': 'http://www.panda.tv/10091',
+        'info_dict': {
+            'id': '10091',
+            'display_id': '10091',
+            'ext': 'flv',
+            'title': 're:.+',
+            'description': 're:.+',
+            'uploader': '囚徒',
+            'uploader_id': 3010056,
+            'is_live': True,
+        },
+        'params': {
+            'skip_download': True,
+        },
+    }]
+
+    def _real_extract(self, url):
+        video_id = self._match_id(url)
+
+        room_id = video_id
+
+        config = self._download_json(
+            'http://www.panda.tv/api_room?roomid=%s' % video_id,
+            video_id
+        )
+
+        data = config['data']
+
+        error_code = config.get('errno', 0)
+        if error_code is not 0:
+            error_desc = 'Server reported error %i' % error_code
+            if isinstance(data, (compat_str, compat_basestring)):
+                error_desc += ': ' + data
+            raise ExtractorError(error_desc, expected=True)
+
+        status = data.get('videoinfo').get('status')
+        # 2 = live, 3 = offline
+        if status == '3':
+            raise ExtractorError(
+                'Live stream is offline', expected=True)
+
+        title = data['roominfo']['name']
+        description = data.get('roominfo').get('bulletin')
+        uploader = data.get('hostinfo').get('name')
+        uploader_id = data.get('hostinfo').get('rid')
+        video_info = data.get('videoinfo')
+        plflag = video_info.get('plflag').split('_')[1]
+        room_key = video_info.get('room_key')
+        stream_addr = video_info.get('stream_addr')
+
+        preference = qualities(['HD', 'SD'])
+        formats = []
+        for quality_id, quality in stream_addr.items():
+            if quality == '1':
+                if quality_id == 'SD':
+                    url_format = 'http://pl%s.live.panda.tv/live_panda/%s.flv'
+                elif quality_id == 'HD':
+                    url_format = 'http://pl%s.live.panda.tv/live_panda/%s_mid.flv'
```

Comment:
Please check the now code. Not a skilled python programmer, not sure if I got the point.